### PR TITLE
Guides for Amazon Location Service

### PR DESCRIPTION
**Pull request by jpignata**
*Issue #, if available:* n/a

*Description of changes:* Guides content for the preview of Amazon Location Service.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+</amplify-block>
+</amplify-block-switcher>
+
+This allows guest users access to your backend resources. 
+
+You have now successfully setup and configured your app to use Amazon Location Service!
+
+## Using the Amazon Location Service APIs
+
+In order to access Amazon Location Service APIs you will need to create resources. You can create resources using the [Amazon Location Service console](http://console.aws.amazon.com/location/home) or using the [AWS Command Line Interface (CLI)](https://aws.amazon.com/cli/). Once you have created the resources, you can use these resources by calling the [APIs](https://aws-amplify.github.io/aws-sdk-android/docs/reference/).
```
Comment (richardmcclellan) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:117]: We haven't updated our javadoc (https://aws-amplify.github.io/aws-sdk-android/docs/reference/) since 2.16.12, so it doesn't have any Location API documentation.  I think we either need to remove this link, or get the javadoc up to date (I don't know what the process is for that, yet)

**Review by sammartinez — APPROVED** [review]
Validated JS documentation and made small updates where needed.

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
+## Overview
+
+[Amazon Location Service](http://aws.amazon.com/location) is a geolocation service that lets you add location information to your applications. With Amazon Location Service, you can build applications that provide maps and points of interest, track resources, trigger actions based on location, and convert street addresses into geographic coordinates. This guide provides you with the following information:
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:3]: For a first time reader, who is new to the domain, "assets" may elicit more of a business use case than "resources." My concern with "resources" is that it sounds like an abstract entity within a software system, as opposed to a physical object that a business owns.

```suggestion
[Amazon Location Service](http://aws.amazon.com/location) is a geolocation service that lets you add location information to your applications. With Amazon Location Service, you can build applications that provide maps and points of interest, track assets, trigger actions based on location, and convert street addresses into geographic coordinates. This guide provides you with the following information:
```

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+## Adding authentication to your App using the Amplify CLI
+
+The Amplify Framework uses [Amazon Cognito](https://aws.amazon.com/cognito/) as its primary authentication provider. Amazon Cognito is a robust user directory service that handles user registration, authentication, account recovery, and other operations. Use the following procedure to add authentication to your app using Amazon Cognito and Amplify CLI. 
+
+1. Open a terminal window in Android Studio by clicking Terminal.
+1. Install the [Amplify CLI](https://docs.amplify.aws/cli/start/install) by running the following command:
+
+    ```bash
+    npm install -g @aws-amplify/cli
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:39]: As this is the Android guide, shall we use the new-and-improved procedure, instead? To wit, `npm` had been a nearly non-existent aspect of my Android development life, before working on Amplify.

```suggestion
    curl -sL https://aws-amplify.github.io/amplify-cli/install | bash && $SHELL
```

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    ```console
+    ? Do you want to use the default authentication and security configuration? 
+      `Default configuration`
+    ? How do you want users to be able to sign in? 
+      `Username`
+    ? Do you want to configure advanced settings? 
+      `No, I am done.`
+    ```
+
+1. Run the following command to push your changes to the cloud. When completed, the awsconfiguration.json file will be updated to reference your newly provisioned backend auth resources.
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:65]:
```suggestion
1. Run the following command to push your changes to the cloud. When completed, the `awsconfiguration.json` file will be updated to reference your newly provisioned backend auth resources.
```

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+```java
+AWSMobileClient.getInstance().initialize(getApplicationContext(), new Callback<UserStateDetails>() {
+    @Override
+    public void onResult(UserStateDetails userStateDetails) {
+        Log.i("QuickStart", "onResult: " + userStateDetails.getUserState());
+    }
+
+    @Override
+    public void onError(Exception e) {
+        Log.e("QuickStart", "Initialization error: ", e);
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:88]:
```suggestion
    public void onError(Exception error) {
        Log.e("QuickStart", "Initialization error: ", error);
```

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+<amplify-block name="Kotlin">
+
+```kotlin
+AWSMobileClient.getInstance().initialize(applicationContext, object : Callback<UserStateDetails> {
+    override fun onResult(userStateDetails: UserStateDetails) {
+        Log.i("QuickStart", "onResult: " + userStateDetails.userState)
+    }
+
+    override fun onError(e: Exception) {
+        Log.e("QuickStart", "Initialization error: ", e)
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:103]:
```suggestion
    override fun onError(error: Exception) {
        Log.e("QuickStart", "Initialization error: ", error)
```

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+                "Action": "geo:SearchPlaceIndexForText",
+                "Resource": "[ARN]"
+            }
+        ]
+    }
+    ```
+
+1. Click on the **Review policy** button.
+2. In the **Name** field, enter *LocationTracker*.
+3. Click on the **Create policy** button. You have now successfully added authentication to your Android app.
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:167]:
```suggestion
1. In the **Name** field, enter *LocationTracker*.
1. Click on the **Create policy** button. You have now successfully added authentication to your Android app.
```

File: docs/guides/location-service/fragments/setting-up-your-app-ios.md [hunk 8]
```diff
@@ -0,0 +1,249 @@
+## Overview
+
+[Amazon Location Service](http://aws.amazon.com/location) is a geolocation service that lets you add location information to your applications. With Amazon Location Service, you can build applications that provide maps and points of interest, track resources, trigger actions based on location, and convert street addresses into geographic coordinates. This guide provides you with the following information:
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/setting-up-your-app-ios.md:3]: There seems to be a lot of duplicated content in this fragment. Is it possible to get the English-language content into a common root, and then use templates for the platform-specific code snippets?

File: docs/guides/location-service/fragments/tracking-device-location-android.md [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+    AWSLocationTracker tracker;
+
+    AWSMobileClient.getInstance().initialize(getApplicationContext(), new Callback<UserStateDetails>() {
+      @Override
+        public void onResult(UserStateDetails userStateDetails) {
+            tracker = new AWSLocationTracker("MyTracker", AWSMobileClient.getInstance());
+        }
+
+        @Override
+        public void onError(Exception e) {
```
Comment (jamesonwilliams) [inline docs/guides/location-service/fragments/tracking-device-location-android.md:89]:
```suggestion
        public void onError(Exception error) {
```

**Review by jamesonwilliams — APPROVED** [review]
Just trivial comments. Looks good. :ship:

File: docs/guides/location-service/fragments/tracking-device-location-android.md [hunk 10]
```diff
@@ -0,0 +1,208 @@
...
+                "Effect": "Allow",
+                "Action": "geo:BatchUpdateDevicePosition",
+                "Resource": "[ARN]"
+            }
+        ]
+    }
+    ```
+
+1. Click on the **Review policy** button.
+2. In the **Name** field, enter **LocationTracker**.
```
Comment (anfield27) [inline docs/guides/location-service/fragments/tracking-device-location-android.md:49]: Need to adjust the bullet list numbers here

File: docs/guides/location-service/fragments/setting-up-your-app-ios.md (hunk 8 above)
Comment (jpignata) [reply docs/guides/location-service/fragments/setting-up-your-app-ios.md:3]: :thumbsup: Good suggestion for later.

File: docs/guides/location-service/fragments/setting-up-your-app-android.md (hunk 3 above)
Comment (jpignata) [reply docs/guides/location-service/fragments/setting-up-your-app-android.md:39]: This seems valid to me. Can we make this work for Windows users, too? I think there's a tabbed snippet.

File: docs/guides/location-service/fragments/setting-up-your-app-android.md (hunk 2 above)
Comment (jpignata) [reply docs/guides/location-service/fragments/setting-up-your-app-android.md:3]: I tihnk this wordsmithing is great! I'm going to hold it for now since we went over this quite a bit internally, and we can revisit later.

File: docs/guides/location-service/fragments/setting-up-your-app-android.md (hunk 3 above)
Comment (jpignata) [reply docs/guides/location-service/fragments/setting-up-your-app-android.md:43]: Addressed!

File: docs/guides/location-service/fragments/setting-up-your-app-android.md (hunk 1 above)
Comment (jpignata) [reply docs/guides/location-service/fragments/setting-up-your-app-android.md:141]: @richardmcclellan resolved this 👍 .

File: docs/guides/location-service/fragments/setting-up-your-app-js.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
+[Amazon Location Service](https://aws.amazon.com/location/) is a new geolocation service that lets you add location information to your applications. With Amazon Location Service, you can build applications that provide maps and points of interest. You can also track resources, trigger actions based on location, and convert street addresses into geographic coordinates.  
+
+This tutorial topic outlines how to set up your app to use Amazon Location Service. To learn more about the service, see the [Amazon Location Service Developer Guide](https://docs.aws.amazon.com/location/latest/developerguide/).
+
+## Overview
+
+In this tutorial, we’ll do the following:
+
+* Create a new React app as the base for this tutorial.
+* Use the AWS SDK JavaScript v3 package for the Amazon Location Service SDK calls.
```
Comment (wlee221) [inline docs/guides/location-service/fragments/setting-up-your-app-js.md:10]: This should be v2

File: docs/guides/location-service/fragments/setting-up-your-app-js.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+1. From your projects directory, run the following commands:
+
+```bash
+npx create-react-app amazon-location-service-places
+cd amazon-location-service-places
+```
+
+This creates a new React app in a directory called `amazon-location-service-places` and moves into that new directory.
+
+2. Now that you’re in the root directory of the project, run the app with the npm start command. This command also runs a development server so you can see the output generated by the build by navigating to `http://localhost:3000`.
```
Comment (wlee221) [inline docs/guides/location-service/fragments/setting-up-your-app-js.md:30]:
```suggestion
2. Now that you’re in the root directory of the project, run the app with the `npm start` command. This command also runs a development server so you can see the output generated by the build by navigating to `http://localhost:3000`.
```

File: docs/guides/location-service/fragments/setting-up-your-app-js.md (hunk 11 above)
Comment (wlee221) [inline docs/guides/location-service/fragments/setting-up-your-app-js.md:10]:
```suggestion
* Use the AWS SDK JavaScript v2 package for the Amazon Location Service SDK calls.
```

File: docs/guides/location-service/fragments/setting-up-your-app-js.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+    const credentials = await Auth.currentCredentials();
+    const client = new Location({
+        credentials,
+        region: awsconfig.aws_project_region,
+   });
+   return client;
+}
+```
+
+ You’ve now successfully connected your React app to the Amazon Location Service.
```
Comment (wlee221) [inline docs/guides/location-service/fragments/setting-up-your-app-js.md:87]:
```suggestion
You’ve now successfully connected your React app to the Amazon Location Service.
```

File: docs/guides/location-service/fragments/setting-up-your-app-js.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+
+```javascript
+const params = {
+  IndexName: "MyPlaceIndex",
+  Text: "Indianapolis",
+};
+client.searchPlaceIndexForText(params, (err, data) => {
+  if (err) console.error(err);
+  if (data) console.log(data);
+});
```
Comment (wlee221) [inline docs/guides/location-service/fragments/setting-up-your-app-js.md:154]:
```suggestion
const params = {
    IndexName: "MyPlaceIndex",
    Text: "Indianapolis",
};
client.searchPlaceIndexForText(params, (err, data) => {
    if (err) console.error(err);
    if (data) console.log(data);
});
```

**Review by wlee221 — CHANGES_REQUESTED** [review]
LGTM except some formatting nits

**Review by wlee221 — APPROVED** [review]

**Comment by wlee221** [comment]
Approving so this doesn't block 👍

File: docs/guides/location-service/fragments/setting-up-your-app-js.md (hunk 12 above)
Comment (sammartinez) [reply docs/guides/location-service/fragments/setting-up-your-app-js.md:30]: will adjust

File: docs/guides/location-service/fragments/setting-up-your-app-js.md (hunk 11 above)
Comment (sammartinez) [reply docs/guides/location-service/fragments/setting-up-your-app-js.md:10]: good catch, will adjust

File: docs/guides/location-service/fragments/setting-up-your-app-js.md (hunk 13 above)
Comment (sammartinez) [reply docs/guides/location-service/fragments/setting-up-your-app-js.md:87]: adjusting

File: docs/guides/location-service/fragments/setting-up-your-app-js.md (hunk 14 above)
Comment (sammartinez) [reply docs/guides/location-service/fragments/setting-up-your-app-js.md:154]: I believe these ones are fine.

**Comment by sammartinez** [comment]
> Approving so this doesn't block 👍

Thanks @wlee221 for the callouts! They are up to date now.

File: docs/guides/location-service/fragments/setting-up-your-app-android.md [hunk 15]
```diff
@@ -0,0 +1,259 @@
...
+
+<amplify-block-switcher>
+
+<amplify-block name="Java">
+
+```java
+AWSMobileClient.getInstance().initialize(getApplicationContext(), new Callback<UserStateDetails>() {
+    @Override
+    public void onResult(UserStateDetails userStateDetails) {
+        ExecutorService executorService = Executors.newFixedThreadPool(4);
```
Comment (drochetti) [inline docs/guides/location-service/fragments/setting-up-your-app-android.md:205]: question, why is this needed and why 4?

File: docs/guides/location-service/fragments/setting-up-your-app-ios.md [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+func searchForPlaces() {
+    let request = AWSLocationSearchPlaceIndexForTextRequest()!
+    request.text = "dessert"
+    request.indexName = "MyPlaceIndex"
+    let result = AWSLocation.default().searchPlaceIndex(forText: request)
+    result.continueWith { (task) -> Any? in
+        if let error = task.error {
+            print("error \(error)")
+        } else if let taskResult = task.result {
+            print("taskResult \(taskResult)")
```
Comment (drochetti) [inline docs/guides/location-service/fragments/setting-up-your-app-ios.md:271]: maybe not a blocker, but we could add a couple of examples on what to do with the `result` (display place name or any other common property a customer could use)

File: docs/guides/location-service/fragments/setting-up-your-app-ios.md [hunk 17]
```diff
@@ -0,0 +1,276 @@
...
+
+Here is how you can search for places using the place index you just created:
+
+```swift
+// Add the following import to the top of your class
+import AWSLocation
+
+func searchForPlaces() {
+    let request = AWSLocationSearchPlaceIndexForTextRequest()!
+    request.text = "dessert"
```
Comment (drochetti) [inline docs/guides/location-service/fragments/setting-up-your-app-ios.md:264]:
```suggestion
    request.text = "Space Needle"
```

File: docs/guides/location-service/fragments/setting-up-your-app-android.md (hunk 15 above)
Comment (jamesonwilliams) [reply docs/guides/location-service/fragments/setting-up-your-app-android.md:205]: Previously, we had a construct like:
```java
new Thread(() -> {
    // do location stuff ...
}).start();
```
Android's official documentation [suggested doing background tasks on an `Executor`](https://developer.android.com/guide/background/threading#executing-in-a-background-thread), and chooses a `Executors.newFixedThreadPool(4)` somewhat at random.

I think the goal here was to show some background work in a way that was close to the platforms docs.

**Review by jamesonwilliams — APPROVED** [review]

**Review by sammartinez — APPROVED** [review]
LGTM 🌮  Can confirm works as expected and removal of iOS doc
